### PR TITLE
Gas Metering and Rate Limiting 

### PR DESCRIPTION
**Pull request by ben-chain**
## Description
Adds the ability of the EM to meter OVM gas, and adds a rate limiting feature to bound the gas consumed by a rollup chain per unit time.

Note: this is a rebase of #162 on a new branch.

## Todos
- Clean up tests so that `SimpleGas.consumeGas(...)` consumes exactly the gas specified, not slightly more (due to returning)
- Add Full Node integration tests to be sure we are handling errors gracefully

## Questions
- Even failed transactions need to increment the cumulative gas.  This means `allowRevert` cannot really be `true` for the metering to work.  Are we down to remove `allowRevert` altogether?
- We are hitting the stack depth in the EM, so some of the changes are a little ugly.  Should I pull out the gas metering stuff into a wrapper function which `CALL`s the EM once again, like we do for `executeTransaction`, or otherwise implement some library pattern for this logic?
- Any other TODOs pop out?

### Fixes
- Fixes YAS 423

## Contributing Agreement

- [x] I have read and understood the [Optimism Contributing Guide and Code of Conduct](https://github.com/ethereum-optimism/optimism-monorepo/blob/master/.github/CONTRIBUTING.md) and am following those guidelines in this pull request.

### CODE REVIEW

File: packages/contracts/contracts/optimistic-ethereum/utils/libraries/DataTypes.sol [hunk 1]
```diff
@@ -27,11 +27,19 @@ library DataTypes {
...
-        uint gasLimit;
         address ovmActiveContract;
         address ovmMsgSender;
         address ovmTxOrigin;
         address l1MessageSender;
+        uint ovmTxGasLimit;
+    }
+
+    struct GasMeterConfig {
+        uint OvmTxFlatGasFee; // The flat gas fee imposed on all transactions
```
Comment (karlfloersch) [inline packages/contracts/contracts/optimistic-ethereum/utils/libraries/DataTypes.sol:38]: Fixed cost gas fee or something

File: packages/contracts/contracts/optimistic-ethereum/ovm/ExecutionManager.sol [hunk 2]
```diff
@@ -206,21 +227,72 @@ contract ExecutionManager is ContractResolver {
...
+                >
+                gasMeterConfig.MaxQueuedGasPerEpoch
+            ) {
+                emit EOACallRevert("Transaction gas limit exceeds remaining gas for this epoch and queue origin.");
+                assembly {
+                    return(0,0)
+                }
+            }
+        }
+
```
Comment (willmeister) [inline packages/contracts/contracts/optimistic-ethereum/ovm/ExecutionManager.sol:295]: Can this get pulled out into a `validateGasLImit(...)` function? Doesn't look like there are any variables that are set from this that are used in `executeTransaction(...)` so it should be cleanly extractable, right?

File: packages/contracts/contracts/optimistic-ethereum/ovm/ExecutionManager.sol [hunk 3]
```diff
@@ -259,25 +335,60 @@ contract ExecutionManager is ContractResolver {
...
+            ovmCallReturnDataSize := returndatasize
             result := mload(0x40)
             let resultData := add(result, 0x20)
-            returndatacopy(resultData, 0, returndatasize)
+            returndatacopy(resultData, 0, ovmCallReturnDataSize)
+            mstore(result, ovmCallReturnDataSize)
+            mstore(0x40, add(resultData, ovmCallReturnDataSize))
+        }
+
+        // get the consumed by execution itself
```
Comment (willmeister) [inline packages/contracts/contracts/optimistic-ethereum/ovm/ExecutionManager.sol:364]:
```suggestion
        // get the gas consumed by execution itself
```

File: packages/contracts/contracts/optimistic-ethereum/ovm/ExecutionManager.sol [hunk 4]
```diff
@@ -206,21 +227,72 @@ contract ExecutionManager is ContractResolver {
...
+        // check for gas rate limit
+        // TODO: split up EM somehow?  We are at max stack depth so can't use vars here yet
+        // TODO: make queue origin an enum?  or just configure better?
+        if (_queueOrigin == 0) {
+            if (
+                getCumulativeSequencedGas()
+                - getCumulativeSequencedGasAtEpochStart()
+                + _ovmTxGasLimit
+                >
+                gasMeterConfig.MaxSequencedGasPerEpoch
```
Comment (willmeister) [inline packages/contracts/contracts/optimistic-ethereum/ovm/ExecutionManager.sol:274]: I swear our code formatter is just trolling us 🤣

File: packages/contracts/package.json [hunk 5]
```diff
@@ -40,6 +40,7 @@
     "@eth-optimism/core-db": "^0.0.1-alpha.25",
     "@eth-optimism/core-utils": "^0.0.1-alpha.25",
     "@eth-optimism/solc-transpiler": "^0.0.1-alpha.27",
+    "@eth-optimism/rollup-core": "^0.0.1-alpha.28",
```
[the remark this answers is not in the document]
Comment (ben-chain) [reply packages/contracts/package.json]: Yeah, removed--it was working fine locally but the CI didn't like it.

**Pull request merged**